File: scripts/run_floorplan_recon.py
from __future__ import annotations

import argparse
import sys
from dataclasses import asdict
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[1]
BACKEND_DIR = REPO_ROOT / "backend"
sys.path.insert(0, str(BACKEND_DIR))

from sqlalchemy import create_engine  # noqa: E402
from sqlalchemy.orm import sessionmaker  # noqa: E402

from app.models import Base  # noqa: E402
from app.models.transaction import Transaction  # noqa: E402
from app.schemas.reconciliation import ReconciliationException, ReconciliationResponse  # noqa: E402
from app.schemas.transaction import SourceType  # noqa: E402
from app.services.reconciliation_engine import reconcile_transactions  # noqa: E402
from app.services.transaction_normalizer import normalize_transactions_from_csv  # noqa: E402


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Run local-only BOA vs Dealertrack floorplan reconciliation."
    )
    parser.add_argument("--boa-file", required=True, help="Path to BOA billing statement CSV.")
    parser.add_argument(
        "--dealertrack-file",
        required=True,
        help="Path to Dealertrack floorplan CSV. Convert XLSX exports to CSV first.",
    )
    args = parser.parse_args()

    boa_file = Path(args.boa_file).expanduser()
    dealertrack_file = Path(args.dealertrack_file).expanduser()

    try:
        _validate_csv_path(boa_file, "BOA")
        _validate_csv_path(dealertrack_file, "Dealertrack")

        result = run_reconciliation(boa_file=boa_file, dealertrack_file=dealertrack_file)
    except LocalReconError as error:
        print(f"Error: {error}", file=sys.stderr)
        return 1

    print_reconciliation_result(result)
    return 0


def run_reconciliation(boa_file: Path, dealertrack_file: Path) -> ReconciliationResponse:
    engine = create_engine("sqlite:///:memory:")
    session_local = sessionmaker(autocommit=False, autoflush=False, bind=engine)
    Base.metadata.create_all(bind=engine)

    db = session_local()
    try:
        _load_file(db, boa_file, SourceType.boa)
        _load_file(db, dealertrack_file, SourceType.dealertrack)
        return reconcile_transactions(db)
    finally:
        db.close()
        Base.metadata.drop_all(bind=engine)
        engine.dispose()


def print_reconciliation_result(result: ReconciliationResponse) -> None:
    boa_only = _exceptions_by_type(result, "missing_in_dealertrack")
    dealertrack_only = _exceptions_by_type(result, "missing_in_boa")
    duplicate_dealertrack = _exceptions_by_type(result, "duplicate_transaction")

    print(f"matched count: {result.matched_count}")
    print(f"exceptions count: {result.exception_count}")
    print(f"duplicates count: {result.duplicate_count}")
    print()

    print("matches:")
    if result.match_groups:
        for group in result.match_groups:
            boa_transaction = group.transactions[0]
            dealertrack_transaction = group.transactions[1]
            print(
                "  "
                f"{_row_label(boa_transaction)} <-> {_row_label(dealertrack_transaction)} | "
                f"reason={group.match_reason} | confidence={group.confidence_score:.2f}"
            )
    else:
        print("  none")

    print()
    _print_exception_section("BOA-only rows", boa_only)
    _print_exception_section("Dealertrack-only rows", dealertrack_only)
    _print_exception_section("duplicate Dealertrack rows", duplicate_dealertrack)


def _load_file(db, file_path: Path, source_type: SourceType) -> None:
    transactions, validation_errors = normalize_transactions_from_csv(
        content=file_path.read_bytes(),
        source_type=source_type,
    )

    if validation_errors:
        formatted_errors = "; ".join(
            f"row {error.get('row')}: {error.get('message')}" for error in validation_errors
        )
        raise LocalReconError(f"{source_type.value} file has validation errors: {formatted_errors}")

    db.add_all(Transaction(**asdict(transaction)) for transaction in transactions)
    db.commit()


def _validate_csv_path(file_path: Path, label: str) -> None:
    if not file_path.exists():
        raise LocalReconError(f"{label} file does not exist: {file_path}")
    if file_path.suffix.lower() in {".xlsx", ".xls"}:
        raise LocalReconError(
            f"{label} file is an Excel workbook. Convert it to CSV first; see "
            "scripts/convert_xlsx_to_csv.md."
        )
    if file_path.suffix.lower() != ".csv":
        raise LocalReconError(f"{label} file must be a CSV file: {file_path}")


def _exceptions_by_type(
    result: ReconciliationResponse,
    exception_type: str,
) -> list[ReconciliationException]:
    return [
        exception for exception in result.exceptions if exception.exception_type == exception_type
    ]


def _print_exception_section(title: str, exceptions: list[ReconciliationException]) -> None:
    print(f"{title}: {len(exceptions)}")
    if not exceptions:
        print("  none")
        print()
        return

    for exception in exceptions:
        print(f"  {_row_label(exception.transaction)} | {exception.description}")
    print()


def _row_label(transaction) -> str:
    return (
        f"{transaction.source_type.value.upper()} "
        f"id={transaction.id} "
        f"stock={transaction.stock_number or 'n/a'} "
        f"vin={transaction.vin or 'n/a'} "
        f"ref={transaction.reference_number or 'n/a'} "
        f"amount={transaction.amount}"
    )


class LocalReconError(Exception):
    pass


if __name__ == "__main__":
    raise SystemExit(main())

File: server/src/cli/runFloorplanRecon.ts
import { readFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import { resolve } from "node:path";

import type {
  ReconciliationException,
  ReconciliationResponse,
  TransactionSummary,
} from "../domain/types.js";
import { MemoryTransactionRepository } from "../repositories/transactionRepository.js";
import { reconcileTransactions } from "../services/reconciliationEngine.js";
import { normalizeTransactionsFromCsv } from "../services/transactionNormalizer.js";

type CliArgs = {
  boaFile: string;
  dealertrackFile: string;
};

export async function runLocalFloorplanRecon(args: CliArgs): Promise<ReconciliationResponse> {
  validateCsvPath(args.boaFile, "BOA");
  validateCsvPath(args.dealertrackFile, "Dealertrack");

  const repository = new MemoryTransactionRepository();
  await loadFile(repository, args.boaFile, "boa");
  await loadFile(repository, args.dealertrackFile, "dealertrack");
  return reconcileTransactions(repository);
}

export function formatReconciliationResult(result: ReconciliationResponse): string {
  const statementNotOnGl = exceptionsByPlacement(result, "statement");
  const scheduleNotOnStatement = exceptionsByPlacement(result, "schedule");
  const manualReview = exceptionsByPlacement(result, "manual_review");
  const lines: string[] = [];

  lines.push(`matched count: ${result.matched_count}`);
  lines.push(`exceptions count: ${result.exception_count}`);
  lines.push(`duplicates count: ${result.duplicate_count}`);
  lines.push("");
  lines.push("matches:");

  if (result.match_groups.length === 0) {
    lines.push("  none");
  } else {
    for (const group of result.match_groups) {
      const boaTransaction = group.transactions[0];
      const dealertrackTransaction = group.transactions[1];
      lines.push(
        `  ${rowLabel(boaTransaction)} <-> ${rowLabel(dealertrackTransaction)} | reason=${group.match_reason} | confidence=${group.confidence_score.toFixed(
          2,
        )}`,
      );
    }
  }

  lines.push("");
  appendExceptionSection(lines, "On statement-not on GL", statementNotOnGl);
  appendExceptionSection(lines, "On schedule-not on statement", scheduleNotOnStatement);
  appendExceptionSection(lines, "Needs manual review", manualReview);
  appendVinPresenceDiagnosticSection(lines, result);

  return lines.join("\n");
}

async function loadFile(
  repository: MemoryTransactionRepository,
  filePath: string,
  sourceType: "boa" | "dealertrack",
) {
  const content = await readFile(filePath);
  const result = normalizeTransactionsFromCsv(content, sourceType);

  if (result.validationErrors.length > 0) {
    const formattedErrors = result.validationErrors
      .map((error) => `row ${error.row}: ${error.message}`)
      .join("; ");
    throw new LocalReconError(`${sourceType} file has validation errors: ${formattedErrors}`);
  }

  await repository.insertMany(result.transactions);
}

function validateCsvPath(filePath: string, label: string): void {
  if (!existsSync(filePath)) {
    throw new LocalReconError(`${label} file does not exist: ${filePath}`);
  }
  if (/\.(xlsx|xls)$/i.test(filePath)) {
    throw new LocalReconError(
      `${label} file is an Excel workbook. Convert it to CSV first; see scripts/convert_xlsx_to_csv.md.`,
    );
  }
  if (!/\.csv$/i.test(filePath)) {
    throw new LocalReconError(`${label} file must be a CSV file: ${filePath}`);
  }
}

function exceptionsByPlacement(
  result: ReconciliationResponse,
  placement: "statement" | "schedule" | "manual_review",
): ReconciliationException[] {
  return result.exceptions.filter((exception) => exceptionPlacement(exception) === placement);
}

function exceptionPlacement(exception: ReconciliationException): "statement" | "schedule" | "manual_review" {
  if (
    exception.exception_type === "needs_review_vin6_only" ||
    exception.exception_category === "vin6_match_amount_mismatch"
  ) {
    return "manual_review";
  }
  if (exception.exception_type === "missing_in_dealertrack" || exception.source_type === "boa") {
    return "statement";
  }
  return "schedule";
}

function appendExceptionSection(
  lines: string[],
  title: string,
  exceptions: ReconciliationException[],
): void {
  lines.push(`${title}: ${exceptions.length}`);
  if (exceptions.length === 0) {
    lines.push("  none");
    lines.push("");
    return;
  }

  for (const exception of exceptions) {
    lines.push(`  ${rowLabel(exception.transaction)} | ${neutralExceptionPrompt(exception)}`);
  }
  lines.push("");
}

function neutralExceptionPrompt(exception: ReconciliationException): string {
  const placement = exceptionPlacement(exception);
  if (placement === "statement") {
    return "BOA statement row with no matching Dealertrack/GL row";
  }
  if (placement === "schedule") {
    return "Dealertrack/GL row with no matching BOA statement row";
  }
  return "VIN appears on both sides but amount differs; review manually";
}

function appendVinPresenceDiagnosticSection(
  lines: string[],
  result: ReconciliationResponse,
): void {
  const diagnostics = result.vin_presence_diagnostics;
  lines.push("VIN presence diagnostics:");
  lines.push(`  BOA extracted VINs: ${diagnostics.extracted_vin_sets.boa.length}`);
  lines.push(`  Dealertrack extracted VINs: ${diagnostics.extracted_vin_sets.dealertrack.length}`);
  appendVinList(
    lines,
    "  VINs in Dealertrack but not BOA",
    diagnostics.vin_presence_exceptions.dealertrack_not_in_boa,
  );
  appendVinList(
    lines,
    "  VINs in BOA but not Dealertrack",
    diagnostics.vin_presence_exceptions.boa_not_in_dealertrack,
  );
  lines.push(
    `  VINs present in both but transaction-unmatched: ${diagnostics.transaction_unmatched_shared_vins.length}`,
  );
  for (const entry of diagnostics.transaction_unmatched_shared_vins) {
    lines.push(
      `    ${entry.vin} | reason=${entry.likely_reason} | boa_ids=${entry.unmatched_boa_transaction_ids.join(
        "/",
      ) || "none"} | dealertrack_ids=${entry.unmatched_dealertrack_transaction_ids.join("/") || "none"}`,
    );
  }
  lines.push("");
}

function appendVinList(lines: string[], title: string, vins: string[]): void {
  lines.push(`${title}: ${vins.length}`);
  for (const vin of vins) {
    lines.push(`    ${vin}`);
  }
}

function rowLabel(transaction: TransactionSummary): string {
  return `${transaction.source_type.toUpperCase()} id=${transaction.id} stock=${
    transaction.stock_number ?? "n/a"
  } vin=${transaction.vin ?? "n/a"} ref=${transaction.reference_number ?? "n/a"} amount=${
    transaction.amount
  }`;
}

function parseArgs(argv: string[]): CliArgs {
  const args = new Map<string, string>();
  for (let index = 0; index < argv.length; index += 2) {
    args.set(argv[index], argv[index + 1]);
  }

  const boaFile = args.get("--boa-file");
  const dealertrackFile = args.get("--dealertrack-file");
  if (!boaFile || !dealertrackFile) {
    throw new LocalReconError("Usage: npm run recon -- --boa-file <boa.csv> --dealertrack-file <dealertrack.csv>");
  }

  return {
    boaFile: resolve(boaFile),
    dealertrackFile: resolve(dealertrackFile),
  };
}

class LocalReconError extends Error {}

if (import.meta.url === `file://${process.argv[1]}`) {
  runLocalFloorplanRecon(parseArgs(process.argv.slice(2)))
    .then((result) => {
      console.log(formatReconciliationResult(result));
    })
    .catch((error) => {
      console.error(`Error: ${error instanceof Error ? error.message : String(error)}`);
      process.exitCode = 1;
    });
}
